osu.Game/Database/LegacyModelExporter.cs

```diff
@@ -0,0 +1,164 @@
+        private readonly List<TModel> exportingModels = new List<TModel>();
+
+        /// <summary>
+        /// Construct exporter.
+        /// Create a new exporter for each export, otherwise it will cause confusing notifications.
+        /// </summary>
+        /// <param name="storage">Storage for storing exported files. Basically it is used to provide export stream</param>
+        /// <param name="realm">The RealmAccess used to provide the exported file.</param>
+        protected LegacyModelExporter(Storage storage, RealmAccess realm)
+        {
+            exportStorage = storage.GetStorageForDirectory(@"exports");
+            UserFileStorage = storage.GetStorageForDirectory(@"files");
+            realmAccess = realm;
+        }
+
+        /// <summary>
+        /// Export the model to default folder.
+        /// </summary>
+        /// <param name="model">The model should export.</param>
+        /// <param name="cancellationToken">
+        /// The Cancellation token that can cancel the exporting.
+        /// If specified CancellationToken, then use it. Otherwise use PostNotification's CancellationToken.
+        /// </param>
+        /// <returns></returns>
+        public async Task ExportAsync(TModel model, CancellationToken? cancellationToken = null)
+        {
+            // check if the model is being exporting already
+            if (!exportingModels.Contains(model))
+            {
+                exportingModels.Add(model);
+            }
+            else
+            {
+                PostNotification?.Invoke(new SimpleErrorNotification
+                {
+                    Text = "File is being exported"
+                });
+                return;
+            }
+
+            string itemFilename = GetFilename(model).GetValidFilename();
+            IEnumerable<string> existingExports =
+                exportStorage
+                    .GetFiles(string.Empty, $"{itemFilename}*{FileExtension}")
+                    .Concat(exportStorage.GetDirectories(string.Empty));
+            string filename = NamingUtils.GetNextBestFilename(existingExports, $"{itemFilename}{FileExtension}");
+            bool success = false;
+
+            ProgressNotification notification = new ProgressNotification
+            {
+                State = ProgressNotificationState.Active,
+                Text = "Exporting...",
+                CompletionText = "Export completed"
+            };
+            notification.CompletionClickAction += () => exportStorage.PresentFileExternally(filename);
+            PostNotification?.Invoke(notification);
+
+            try
+            {
+                using (var stream = exportStorage.CreateFileSafely(filename))
+                {
+                    success = await ExportToStreamAsync(model, stream, notification, cancellationToken ?? notification.CancellationToken).ConfigureAwait(false);
+                }
+            }
+            catch (OperationCanceledException)
```

Comment:
We probably want to look into the best practices of catching `OperationCanceledException`. I'm pretty sure we don't want to catch without a `rethrow`, but we're already doing this (likely incorrectly) in other places.